minetest/minetest
# Complete collision rewrite

This PR is for a rewrite of the collision code using an inverted index data structure.

The goal is to provide reliable collision detection for mobs. The current collision implementation makes a list of collision boxes near where the mob is moving, and looks through the whole list repeatedly, trying to check if the mob will collide with any of them.

The new data structure efficiently identifies the collision boxes in the order they collide with a moving query box. Each dimension of the boxes are indexed, and the indexes are placed into a heap so that the algorithm can track when boxes pass into or out of the space of the query box in one dimension. When a box already overlaps in 2 dimensions and crosses over into the 3rd detection, a collision is registered. To detect if the query box is supported (standing on ground, stepping up against a wall), look for boxes that are adjacent to the required face and overlap in the other two dimensions.

There are many reported problems with entities failing to properly detect collisions while moving. This also resolves issues with collision unit tests.

## To do

This PR is a Work in Progress

- [X] There is currently mangling of memory that occasionally fails unit tests, including segmentation faults and values being overwritten in memory. I'm making a PR largely to see what valgrind reports (I am developing on Windows). The offending bug overwrote the most significant 16 bits of a u32, so that narrows the hunt to 16-bit operations, or perhaps freeing memory that was not allocated.
- [ ] Finish writing a query for collision detection with linear motion. (Basic movement.)
- [ ] Write a query for collision detection and loss-of-support on one face detection with linear motion. (Stepping up against a wall, stepping forward over a wall.)
- [ ] Write a query for collision detection and loss-of-support on two faces detection with linear motion. (Stepping up when the entity collided directly into a corner.)
- [X] Write unit tests for the query context classes.
- [ ] Write unit tests for the queries.
- [ ] Rewrite collision detection logic using the queries.
- [ ] Resolve bugs identified by unit tests.
- [ ] Field test extensively.
- [ ] Compare performance characteristics (esp. memory and processor time) with existing implementation.

## How to test

1. Build with unit tests enabled.
2. Run the unit tests. look for failures in the TestInvertedIndex and TestCollisionQuery groups.
3. Run minetest. Build a tall wall around a space. Leave a gap and build another tall wall a few blocks out from the first wall. Spawn 10 mobs that won't kill each other inside the first wall. Sit on top of the wall and wait a while to see if the mobs cross over into the space between the walls.

## Future directions

- If the entity is moving very fast, the number of possible collision boxes is very large, and there may be many collision boxes that share space with the query box in one dimension, even though they are far away in other dimensions. This implementation will have to track them, in case they eventually share space in the other dimensions. Depending on how this works in practice it may be beneficial to separately index each block, and only attend to boxes that are in an active block.  Incidentally, the number of collision boxes in an index would be much smaller, so that the memory used for each could be cut in half. 

- When there are multiple entities moving in an area, they could all share the same inverted indices.

- Explore whether this implementation could be extended to detect entities that collide during movement instead of treating the other entities as immobile.

- If the indexes could be updated efficiently when nodes change, the data structure for each active block could be kept in memory for reuse rather than rebuilding it repeatedly. Or, they could be cached if there were a way to invalidate them when a block changes.

- The new implementation would be amenable to incorporating more physics during query execution (friction, spring coefficients, continuous acceleration). Performance overhead is probably reasonable.
